man/alloc.3
.TH qlibs: alloc 3
.SH NAME
alloc \- allocate memory
.SH SYNTAX
.B #include \(dqalloc.h\(dq

char *\fBalloc\fP(\fInew\fR);

void \fBalloc_free\fP(\fIx\fR);

void \fBalloc_re\fP(&\fIx\fR,\fIold\fR,\fInew\fR);

char *\fIx\fR;
.br
unsigned int \fIold\fR;
.br
unsigned int \fInew\fR;
.SH DESCRIPTION
.B alloc
allocates enough space from the heap for
.I new
bytes of data,
adequately aligned for any data type.
.I new
may be 0.
.B alloc
returns a pointer to the space.
If space is not available,
.B alloc
returns 0,
setting
.B errno
appropriately.

.B alloc_free
returns space to the heap.

.B alloc_re
expands the space allocated to
.I x
from
.I old
bytes to
.I new
bytes.
It allocates new space,
copies
.I old
bytes from the old space to the new space,
returns the old space to the heap,
and changes
.I x
to point to the new space.
It then returns 1.
If space is not available,
.B alloc_re
returns 0,
leaving the old space alone.
.SH "CVE-2005-1513"
This version of
.B alloc
respects
.I limits.h
to avoid memory resource exhaustion.
.SH "SEE ALSO"
sbrk(2),
malloc(3),
error(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
		DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
		DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.
fehQlibs-15: 	Added the Guninski patch for alloc from Qualys (CVE-2005-1513).
		dns_ipq (IP qualify) now with unified IPv4/IPv6 handling and evaluation 
		of DNS well-know names 'localhost' and 'ipv[4|6]-loopback' (RFC 6761).
		Fixed DNS lookup for IPv6 addresses. Adjusted DNS man pages.
fehQlibs-15a:  	Again, changes for dns_ipq.c concerning return codes.
		Should have only impact in case using DJB's qualification format.
		Documentation and man page fixes for DNS stub resolver. 
		GCC 10 linker warning for external 'ipv4socket' solved.
fehQlibs-15b:	GCC 10 compiler issues solved. DNS TLSA record correct value.
fehQlibs-16:	19961114 change: dns_ip() now recognizes [1.2.3.4] and [f80::a]. tnx DS.
		scan_long() included for mess822x support.
		constmap() added from qmail and written man page.
		Updated getln man page to refer buffer.
		Changed ipv4/6-localhost and ipv4/6-loopback to ip4/6-localhost and ip4/6-loopback.
fehQlibs-16+:	GCC 10.2 conformance; ipv4socket definitition removed in socket_send()
		and socket_connect().
		This is a major desaster: 'external' definitions depend on time-of-usage.
fehQlibs-17:	Removed dependency on ipv4socket entirely. 
		Reworked socket interface + doc. Removed obsolete dns_sortip() function.
		genalloc.h is now separate. Installation of man pages described.
fehQlibs-18
		Due to problems binding with IPv4-mapped IPv6 addresses for a DNS lookup,
		splitted up randombind into randombind4 and randombind6.
		socket operations on IPv4 use IPv4 address only. (tx. Kouichi).
fehQlibs-19
		Provided socket_accept4 again to avoid IPv4-mapped IPv6 address binding problems.
fehQlibs-20
		Added uint8p.c functions. Fixed dns_txt.c for multiple labels in RDATA section.
fehQlibs-21
		Fixed wrong return code for DNS_COM (tx Franz). 
		Fixed header and man page for env functions and included fd_coe in man fd.
fehQlibs-22
		Changed ipX_bytestring to return correctly the number of bytes processed.
fehQlibs-22a
		dns_ip A lookup recognizes now 'composit' ip/fqdn correctly; ie. 1.2.3.4.example.com.
fehQlibs-23
		More documentation for dns resolver routines.
		Added function stralloc_free() which was alreday in the header.


alloc.c
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include "byte.h"
#include "alloc.h"

#define ALIGNMENT 16 /* XXX: assuming that this alignment is enough */
#define SPACE 4096   /* must be multiple of ALIGNMENT */

typedef union { char irrelevant[ALIGNMENT]; double d; } aligned;
static aligned realspace[SPACE / ALIGNMENT];
#define space ((char *) realspace)
static unsigned int avail = SPACE; /* multiple of ALIGNMENT; 0<=avail<=SPACE */

/*@null@*//*@out@*/char *alloc(unsigned int n) {
  char *x;

/* Guninski exploit + patch from Qualys (CVE-2005-1513) */

  if (n >= (INT_MAX >> 3)) {
    errno = ENOMEM;
    return 0;
  }

  n = ALIGNMENT + n - (n & (ALIGNMENT - 1)); /* XXX: could overflow */
  if (n <= avail) { avail -= n; return space + avail; }
  x = malloc(n);
  if (!x) errno = ENOMEM;
  return x;
}

void alloc_free(char *x) 
{
  if (x >= space)
    if (x < space + SPACE)
      return; /* XXX: assuming that pointers are flat */
  free(x);
}

int alloc_re(char **x,unsigned int m,unsigned int n) 
{
  char *y;

  y = alloc(n);
  if (!y) return 0;
  byte_copy(y,m,*x);
  qfree(*x);
  *x = y;
  return 1;
}
